be/src/util/frame_of_reference_coding.cpp

```diff
@@ -0,0 +1,261 @@
+    _buffered_values_num += count;
+    in_data += count;
+    return in_data;
+}
+
+template<typename T>
+void ForEncoder<T>::put_batch(const T *in_data, size_t count) {
+    if (_buffered_values_num + count < ForCoding::FRAME_VALUE_NUM) {
+        copy_value(in_data, count);
+        _count += count;
+        return;
+    }
+
+    // 1. padding one frame
+    size_t padding_num = ForCoding::FRAME_VALUE_NUM - _buffered_values_num;
+    in_data = copy_value(in_data, padding_num);
+    bit_packing_one_frame_value(_buffered_values);
+
+    // 2. process frame by frame
+    size_t frame_size = (count - padding_num) / ForCoding::FRAME_VALUE_NUM;
+    for(size_t i = 0; i < frame_size; i ++) {
+        // directly encode value to the bit_writer, don't buffer the value
+        _buffered_values_num = ForCoding::FRAME_VALUE_NUM;
+        bit_packing_one_frame_value(in_data);
+        in_data += ForCoding::FRAME_VALUE_NUM;
+    }
+
+    // 3. process remaining value
+    size_t remaining_num = (count - padding_num) % ForCoding::FRAME_VALUE_NUM;
+    if (remaining_num > 0) {
+        copy_value(in_data, remaining_num);
+    }
+
+    _count += count;
+}
+
+template<typename T>
+void ForEncoder<T>::bit_packing_one_frame_value(const T* input) {
+    _frame_offsets.push_back(_bit_writer.bits_written());
+
+    T min = input[0];
+    T max = input[0];
+    for (uint8_t i = 1; i < _buffered_values_num; ++i) {
+        if (input[i] < min) {
+            min = input[i];
+            continue;
+        }
+
+        if (input[i] > max) {
+            max = input[i];
+        }
+    }
+
+    uint32_t bit_width = 0;
+    if (sizeof(T) == 8) {
+        bit_width = bits64(static_cast<T>(max - min));
+    } else {
+        bit_width = bits(static_cast<T>(max - min));
+    }
+
+    _bit_writer.PutValue(min, sizeof(T) * ForCoding::BYTE_SIZE);
+    for (uint8_t i = 0; i < _buffered_values_num; ++i) {
+        _bit_writer.PutValue(input[i] - min, bit_width);
+    }
+
+    _frame_count++;
+    _buffered_values_num = 0;
+}
+
+template<typename T>
+uint32_t ForEncoder<T>::flush() {
+    uint8_t last_frame_values_num = 0;
+    if (_buffered_values_num != 0) {
+        last_frame_values_num = _buffered_values_num;
+        bit_packing_one_frame_value(_buffered_values);
+    }
+
+    // in order to compute the last frame bit_width when decode frame
+    _frame_offsets.push_back(_bit_writer.bits_written());
+
+    // align to bytes
+    _bit_writer.Flush(true);
+
+    // write the footer
+    for(auto i: _frame_offsets) {
+        _bit_writer.PutValue(i, ForCoding::INT32_SIZE);
+    }
+
+    if (last_frame_values_num != 0) {
+        _bit_writer.PutValue(last_frame_values_num, ForCoding::BYTE_SIZE);
+    }
+
+    _bit_writer.PutValue(_frame_count, ForCoding::INT32_SIZE);
+    _bit_writer.PutValue(_count, ForCoding::INT32_SIZE);
+
+    _bit_writer.Flush();
+    return _bit_writer.bytes_written();
+}
+
+
+template<typename T>
+bool ForDecoder<T>::init() {
```

Comment:
the return type should be `Status` to include error type and message

Reply:
I think return bool is enough in ForDecoder，FrameOfReferencePageDecoder will return Status.